### PR TITLE
[TritonGEN] Use OCL builtins for subgroup block read/write

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Use `intel_sub_group_block_[read|write]` defined in https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups.html, https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_char.html, https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_short.html, https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_long.html, and https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_subgroup_local_block_io.asciidoc.

### CODE REVIEW

File: third_party/intel/lib/Utils/Mangling.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@ std::string getTypeMangling(Type ty, bool isUnsigned) {
       .Case([isUnsigned](IntegerType ty) -> std::string {
         switch (ty.getWidth()) {
         case 8:
-          return "c";
+          return isUnsigned ? "c" : "h";
```
Comment (victor-eds) [inline third_party/intel/lib/Utils/Mangling.cpp:23]:
```suggestion
          return isUnsigned ? "h" : "c";
```

`c` is for char, `h` for unsigned char (see https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-builtin)

File: third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td [hunk 2]
```diff
@@ -397,7 +397,7 @@ def TritonGEN_Matrix2DBlockPrefetchOp : TritonGEN_Op<"2Dblockprefetch">,
 }
 
 def TritonGEN_SIMDBlockReadOp: TritonGEN_Op<"simdblockread">,
-  Results<(outs FixedVectorOf<[TritonGEN_MatrixElemType]>:$res)>,
+  Results<(outs FixedVectorOf<[AnyTypeOf<[AnyI8, AnyI16, AnyI32, AnyI64]>]>:$res)>,
```
Comment (victor-eds) [inline third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td:400]: Can we add float types and wider vectors and abstract in conversion to LLVM?

File: third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td [hunk 3]
```diff
@@ -418,7 +418,7 @@ def TritonGEN_SIMDBlockReadOp: TritonGEN_Op<"simdblockread">,
 def TritonGEN_SIMDBlockWriteOp : TritonGEN_Op<"simdblockwrite">,
   Arguments<(ins
     Arg<LLVM_AnyPointer, "", [MemWrite]>:$ptr,
-    FixedVectorOf<[TritonGEN_MatrixElemType]>:$val
+    FixedVectorOf<[AnyTypeOf<[AnyI8, AnyI16, AnyI32, AnyI64]>]>:$val
```
Comment (victor-eds) [inline third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td:421]: Same

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 4]
```diff
@@ -1254,6 +1254,18 @@ struct TritonMatrix2DBlockPrefetchLowering
...
 };
 
+static bool isTySIMDOCLBuiltinAvailable(VectorType vecTy) {
+  unsigned numElems = vecTy.getNumElements();
+  if (numElems == 1 || numElems == 2 || numElems == 4 || numElems == 8)
+    return true;
+
+  IntegerType elemTy = cast<IntegerType>(vecTy.getElementType());
+  if ((elemTy.getWidth() == 8 || elemTy.getWidth() == 16) && numElems == 16)
+    return true;
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1264]: I don't see `ushort16` supported [here](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_subgroup_local_block_io.asciidoc)

File: test/Conversion/intel/tritongpu_to_llvm_intel_advanced_path.mlir [hunk 5]
```diff
@@ -281,11 +281,12 @@ module attributes {"triton_gpu.num-warps" = 4 : i32, "triton_gpu.threads-per-war
...
-// CHECK:           %[[VAL_10:.*]] = llvm.mul %[[VAL_6]], %[[VAL_5]] : i64
-// CHECK:           %[[VAL_11:.*]] = llvm.getelementptr inbounds %[[VAL_9]]{{\[}}%[[VAL_10]]] : (!llvm.ptr<3>, i64) -> !llvm.ptr<3>, f32
-// CHECK:           %[[VAL_12:.*]] = llvm.load %[[VAL_11]] : !llvm.ptr<3> -> vector<16xf32>
-// CHECK:           llvm.return %[[VAL_12]] : vector<16xf32>
+// CHECK:           %[[VAL_10:.*]] = llvm.bitcast %[[VAL_1]] : vector<16xf32> to vector<16xi32>
+// CHECK:           llvm.call spir_funccc @llvm.genx.GenISA.simdBlockWrite(%[[VAL_9]], %[[VAL_10]]) {{{.*}}} : (!llvm.ptr<3>, vector<16xi32>) -> ()
+// CHECK:           %[[VAL_11:.*]] = llvm.mul %[[VAL_6]], %[[VAL_5]] : i64
+// CHECK:           %[[VAL_12:.*]] = llvm.getelementptr inbounds %[[VAL_9]]{{\[}}%[[VAL_11]]] : (!llvm.ptr<3>, i64) -> !llvm.ptr<3>, f32
+// CHECK:           %[[VAL_13:.*]] = llvm.load %[[VAL_12]] : !llvm.ptr<3> -> vector<16xf32>
+// CHECK:           llvm.return %[[VAL_13]] : vector<16xf32>
```
Comment (victor-eds) [inline test/Conversion/intel/tritongpu_to_llvm_intel_advanced_path.mlir:289]: Isn't this a legal type for the intrinsic? Using approach I mention above we could avoid this bitcast.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 6]
```diff
@@ -1266,7 +1278,16 @@ struct TritonSIMDBlockReadLowering
     VectorType vecTy = op.getRes().getType();
 
     // TODO: Remove GenISA lowering after PoC productization is completed.
-    const StringLiteral funcName = "llvm.genx.GenISA.simdBlockRead";
+    std::string funcName = "llvm.genx.GenISA.simdBlockRead";
+    if (isTySIMDOCLBuiltinAvailable(vecTy)) {
+      unsigned numElems = vecTy.getNumElements();
+      funcName = "intel_sub_group_block_read_u" +
+                 intel::getTypeMangling(vecTy.getElementType(), false) +
+                 (numElems == 1 ? "" : std::to_string(numElems));
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1286]: Can you add a comment mentioning OCL builtin name here differs from regular mangling?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 7]
```diff
@@ -1296,7 +1317,18 @@ struct TritonSIMDBlockWriteLowering
...
+    if (isTySIMDOCLBuiltinAvailable(vecTy)) {
+      unsigned numElems = vecTy.getNumElements();
+      funcName = "intel_sub_group_block_write_u" +
+                 intel::getTypeMangling(vecTy.getElementType(), false) +
+                 (numElems == 1 ? "" : std::to_string(numElems));
+      funcName = "_Z" + std::to_string(funcName.size()) + funcName + "PU3AS" +
+                 std::to_string(ptrTy.getAddressSpace()) +
+                 intel::getTypeMangling(vecTy.getElementType(), true) +
+                 intel::getTypeMangling(vecTy, true);
+    }
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1330]: Can we move this and above to a function to share code?

**Review by victor-eds — COMMENTED** [review]
We can emulate wider vectors with several calls and avoid relying on GenISA intrinsics:
```
%res = triton_gen.simdblockread %ptr : (!llvm.ptr) -> vector<32xi32>
```
is equivalent to:
```
%0 = triton_gen.simdblockread %ptr : (!llvm.ptr) -> vector<8xi32>
%ptr1 = llvm.getelementptr inbounds %ptr[128]
%1 = triton_gen.simdblockread %ptr1 : (!llvm.ptr) -> vector<8xi32>
%ptr2 = llvm.getelementptr inbounds %ptr1[128]
%2 = triton_gen.simdblockread %ptr2 : (!llvm.ptr) -> vector<8xi32>
%ptr3 = llvm.getelementptr inbounds %ptr2[128]
%3 = triton_gen.simdblockread %ptr3 : (!llvm.ptr) -> vector<8xi32>
%res = // Vector concatenation %0 %1 %2 %3
```

**Comment by whitneywhtsang** [comment]
> We can emulate wider vectors with several calls and avoid relying on GenISA intrinsics:
> 
> ```
> %res = triton_gen.simdblockread %ptr : (!llvm.ptr) -> vector<32xi32>
> ```
> 
> is equivalent to:
> 
> ```
> %0 = triton_gen.simdblockread %ptr : (!llvm.ptr) -> vector<8xi32>
> %ptr1 = llvm.getelementptr inbounds %ptr[16] : (!llvm.ptr) -> !llvm.ptr, vector<8xi32>
> %1 = triton_gen.simdblockread %ptr1 : (!llvm.ptr) -> vector<8xi32>
> %ptr2 = llvm.getelementptr inbounds %ptr1[16] : (!llvm.ptr) -> !llvm.ptr, vector<8xi32>
> %2 = triton_gen.simdblockread %ptr2 : (!llvm.ptr) -> vector<8xi32>
> %ptr3 = llvm.getelementptr inbounds %ptr2[16] : (!llvm.ptr) -> !llvm.ptr, vector<8xi32>
> %3 = triton_gen.simdblockread %ptr3 : (!llvm.ptr) -> vector<8xi32>
> %res = // Vector concatenation %0 %1 %2 %3
> ```

Right, I have a local change for that, would like to do that in a separate PR.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 4 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1264]: It is supported https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_short.html starting version 1.1, [here](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_subgroup_local_block_io.asciidoc) likely needs an update.

File: third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td (hunk 2 above)
Comment (whitneywhtsang) [reply third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td:400]: Offline discussion: agreed to have a simple TritonGEN lowering.

File: third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td (hunk 3 above)
Comment (whitneywhtsang) [reply third_party/intel/include/Dialect/TritonGEN/IR/TritonGENOps.td:421]: Offline discussion: agreed to have a simple TritonGEN lowering.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 8]
```diff
@@ -1266,7 +1278,16 @@ struct TritonSIMDBlockReadLowering
     VectorType vecTy = op.getRes().getType();
 
     // TODO: Remove GenISA lowering after PoC productization is completed.
-    const StringLiteral funcName = "llvm.genx.GenISA.simdBlockRead";
+    std::string funcName = "llvm.genx.GenISA.simdBlockRead";
+    if (isTySIMDOCLBuiltinAvailable(vecTy)) {
+      unsigned numElems = vecTy.getNumElements();
+      funcName = "intel_sub_group_block_read_u" +
+                 intel::getTypeMangling(vecTy.getElementType(), false) +
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 6 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 9]
```diff
@@ -1266,7 +1278,16 @@ struct TritonSIMDBlockReadLowering
...
+                 (numElems == 1 ? "" : std::to_string(numElems));
+      funcName = "_Z" + std::to_string(funcName.size()) + funcName + "PU3AS" +
+                 std::to_string(ptrTy.getAddressSpace()) +
+                 intel::getTypeMangling(vecTy.getElementType(), true);
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 7 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: done

File: third_party/intel/lib/Utils/Mangling.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/Utils/Mangling.cpp]: done

**Comment by whitneywhtsang** [comment]
> > > We can emulate wider vectors with several calls and avoid relying on GenISA intrinsics:
> > > ```
> > > %res = triton_gen.simdblockread %ptr : (!llvm.ptr) -> vector<32xi32>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > is equivalent to:
> > > ```
> > > %0 = triton_gen.simdblockread %ptr : (!llvm.ptr) -> vector<8xi32>
> > > %ptr1 = llvm.getelementptr inbounds %ptr[16] : (!llvm.ptr) -> !llvm.ptr, vector<8xi32>
> > > %1 = triton_gen.simdblockread %ptr1 : (!llvm.ptr) -> vector<8xi32>
> > > %ptr2 = llvm.getelementptr inbounds %ptr1[16] : (!llvm.ptr) -> !llvm.ptr, vector<8xi32>
> > > %2 = triton_gen.simdblockread %ptr2 : (!llvm.ptr) -> vector<8xi32>
> > > %ptr3 = llvm.getelementptr inbounds %ptr2[16] : (!llvm.ptr) -> !llvm.ptr, vector<8xi32>
> > > %3 = triton_gen.simdblockread %ptr3 : (!llvm.ptr) -> vector<8xi32>
> > > %res = // Vector concatenation %0 %1 %2 %3
> > > ```
> > 
> > 
> > Right, I have a local change for that, would like to do that in a separate PR.
> 
> How about its performance?

I would like to test its performance, but the SLM path is not working at the moment, so I cannot get a baseline.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 4 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1241]: It fails with https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/lib/SPIRV/OCLUtil.cpp#L1482 in the translator, even though it is supported in https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_short.html.

Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: An internal ticket is created to track the translator assertion, disallow 16xi16 for now.

**Comment by Dewei-Wang-sh** [comment]
anywhere to document what the current ocl built-ins we can use in triton?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 10]
```diff
@@ -1231,6 +1231,40 @@ struct TritonMatrix2DBlockPrefetchLowering
...
+    return true;
+
+  return false;
+}
+
+template <typename OpType, typename = std::enable_if_t<llvm::is_one_of<
+                               OpType, TritonGEN::SIMDBlockReadOp,
+                               TritonGEN::SIMDBlockWriteOp>::value>>
+static std::string getSIMDBlockManglingName(OpType op, VectorType vecTy) {
+  const bool isWrite = std::is_same<OpType, TritonGEN::SIMDBlockWriteOp>::value;
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1251]:
```suggestion
  constexpr bool isWrite = std::is_same_v<OpType, TritonGEN::SIMDBlockWriteOp>;
```

NIT

**Review by victor-eds — APPROVED** [review]
Small NIT. LGTM.

**Comment by whitneywhtsang** [comment]
> anywhere to document what the current ocl built-ins we can use in triton?

We can use any ocl built-ins documented in any OpenCL C extensions, if we want to use an OpenCL C builtin that is not available in any OpenCL C extensions, then we need to send a request to IGC team.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 10 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 11]
```diff
@@ -1231,6 +1231,40 @@ struct TritonMatrix2DBlockPrefetchLowering
...
+  const LLVM::LLVMPointerType ptrTy = op.getPtr().getType();
+  const unsigned numElems = vecTy.getNumElements();
+  // Note: OCL builtin name here differs from regular mangling.
+  std::string funcName = "intel_sub_group_block_";
+  funcName += (isWrite ? "write" : "read");
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 12]
```diff
@@ -1231,6 +1231,40 @@ struct TritonMatrix2DBlockPrefetchLowering
...
+  funcName += "_u" + intel::getTypeMangling(vecTy.getElementType()) +
+              (numElems == 1 ? "" : std::to_string(numElems));
+  funcName =
+      "_Z" + std::to_string(funcName.size()) + funcName + "PU3AS" +
+      std::to_string(ptrTy.getAddressSpace()) +
+      intel::getTypeMangling(vecTy.getElementType(), true /*isUnsigned*/);
+  if (isWrite)
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: done

**Review by etiotto — APPROVED** [review]

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 13]
```diff
@@ -1231,6 +1231,44 @@ struct TritonMatrix2DBlockPrefetchLowering
...
+  else
+    funcName += "read";
+  funcName += "_u" + intel::getTypeMangling(vecTy.getElementType()) +
+              (numElems == 1 ? "" : std::to_string(numElems));
+  funcName =
+      "_Z" + std::to_string(funcName.size()) + funcName + "PU3AS" +
+      std::to_string(ptrTy.getAddressSpace()) +
+      intel::getTypeMangling(vecTy.getElementType(), true /*isUnsigned*/);
+  if constexpr (isWrite)
+    funcName += intel::getTypeMangling(vecTy, true /*isUnsigned*/);
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1268]:
```suggestion
      intel::getTypeMangling(vecTy.getElementType(), /*isUnsigned=*/true);
  if constexpr (isWrite)
    funcName += intel::getTypeMangling(vecTy, /*isUnsigned=*/true);
```
Super NIT. Can merge as is.

**Review by victor-eds — APPROVED** [review]

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 13 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1268]: automerged, changed in https://github.com/intel/intel-xpu-backend-for-triton/pull/2193.

**Pull request merged**